server : use common_token_to_piece instead of common_detokenize

danbev opened the pull request:
TThis commit replaces the call to `common_detokenize` with
`common_token_to_piece` in the populate_token_probs.

The motivation for this change is to avoid an issue where
common_detokenize would remove the word boundary character for tokens,
which caused a regression in the server generated token probabilities.
    
Resolves: https://github.com/ggerganov/llama.cpp/issues/11728

ngxson commented on examples/server/server.cpp:
```diff
@@ -541,12 +541,16 @@ struct completion_token_output {
         json probs_for_token = json::array();
         for (const auto & p : probs) {
-            std::string txt(p.txt);
+            // If the predicted token id is the same as this.tok, then we use the text_to_send instead
```
I'm just wondering if this only happens with `tok == p.tok`, or it also affect other tokens inside `probs_for_token` list

ngxson commented:
Also, a clue is to look around the `common_detokenize`, probably we miss certain params when constructing these detokenized string. The lines around `result.probs.reserve(max_probs);` should be related

CyberShadow commented on examples/server/server.cpp:
```diff
@@ -541,12 +541,16 @@ struct completion_token_output {
+            // of the detokenized token. This is to avoid a mismatch between the text tokens where
+            // the text_to_send token may include a leading whitespace character but the detokenized
+            // token would not.
+            std::string txt = tok == p.tok ? text_to_send : p.txt;
```
Hi, thanks for working on this. Just wanted to mention that fixing the whitespace only for the first token in `top_logprobs` is not going to fix the use case I'm working with - the tokens in the other `top_logprobs` entries need to have the correct whitespace as they did before #10783. Sorry if the issue description was unclear.

danbev replied:
Ah, thanks for the clarification. I'll revert and look into this again now.

CyberShadow reviewed (approved):
I can't comment on the code, but I've tested this patch and it builds successfully + resolves the issue #11728 . Thank you!

ggerganov commented on examples/server/server.cpp:
```diff
@@ -2297,7 +2297,7 @@ struct server_context {
                 result.probs.push_back({
                     cur[i].id,
-                    common_detokenize(ctx, {cur[i].id}, special),
+                    common_detokenize(ctx, {cur[i].id}, special, /* remove_space_prefix */ false),
```
Wouldn't it be simpler to detokenize `{bos, cur[i].id}`?

I would like t avoid adding this extra parameter to the API.

ngxson replied:
@ggerganov maybe I misunderstood, but won't `{bos, cur[i].id}` returns something like `<bos> text`? But what we want here is just ` text`

Also would be nicer if we can query the raw token from vocab, but I don't think we have an API for that.

ggerganov replied:
If `special == false` it won't render the BOS token.

ggerganov replied:
> Also would be nicer if we can query the raw token from vocab, but I don't think we have an API for that.

What about `llama_token_to_piece`?

danbev replied:
@ggerganov @ngxson Thanks, that is a much simpler and better solution!

danbev committed:
server : use common_token_to_piece instead of common_detokenize

This commit replaces the call to common_detokenize with
common_token_to_piece in the populate_token_probs.

The motivation for this change is to avoid an issue where
common_detokenize would remove the word boundary character for tokens,
which caused a regression in the server generated token probabilities.

Resolves: https://github.com/ggerganov/llama.cpp/issues/11728

ngxson commented:
IIRC there is another place having the same logic (near the speculative decode logic), that should be fixed too

danbev committed:
squash! server : use common_token_to_piece instead of common_detokenize

Use common_token_to_piece for post_sampling_probs as well.

danbev commented:
> IIRC there is another place having the same logic (near the speculative decode logic), that should be fixed too

I've taken a look but could not find this in/near the speculative decoding. I did find this [TODO](https://github.com/ggerganov/llama.cpp/blob/b9ab0a4d0b2ed19effec130921d05fb5c30b68c5/examples/server/server.cpp#L3323) though. 

I noticed that this same logic has part of the [post_sampling_probs](https://github.com/ggerganov/llama.cpp/blob/b9ab0a4d0b2ed19effec130921d05fb5c30b68c5/examples/server/server.cpp#L2282) and have updated this in https://github.com/ggerganov/llama.cpp/pull/11740/commits/5deee0a0938986e091fea90a14d77b958c2d77cc.
